### ISSUE TITLE
No output from :Neomake

### ISSUE THREAD

**Issue by adamyonk** [neomake/neomake#493]
Here is my config:
```
autocmd! BufWritePost * Neomake
let g:eslint_path = system('PATH=$(npm bin):$PATH && which eslint | tr -d "\n"')
```
I'm trying to get neomake to run eslint for JS files, but when I do `:Neomake` nothing happens, and when I do `:Neomake!` I see `Neomake: makeprg completed with exit code 2`.

I have no idea how to troubleshoot this, could someone help me through it? Thanks!

**Comment by davidosomething** [comment]
did you steal that from someone's config? e.g. https://github.com/ddelnano/dotfiles/blob/master/vim/plugin_config/neomake.vim

neomake ships pre-configured to use an eslint that was installed globally. There is no config required for eslint other than the optional `autocmd` to run on BufWritePost. (Also `g:eslint_path` is someone's config-specific variable, not a neomake variable, so if you want to use their settings you should take the rest of it too, particularly the part where `let g:neomake_javascript_eslint_exe` is set since _that_ is the neomake variable for setting the executable.)

Easy fix is to just use global eslint
1. `npm install -g eslint` (assuming your node is set up correctly already and can install packages globally via npm, if this fails it is an issue with your node-installation)
1. open a file that has javascript filetype (e.g. anything ending in .js) and save it to trigger that autocmd

**Comment by davidosomething** [comment]
possibly related #323 #296

**Comment by adamyonk** [comment]
OK, yeah I had the wrong var set there, but changing it to

```
autocmd! BufWritePost * Neomake
let g:neomake_javascript_eslint_exe = system('PATH=$(npm bin):$PATH && which eslint | tr -d "\n"')
```

still doesn't fix the exit code 2. I can see that the path to my local projects eslint is correct, but it still doesn't work. Installing eslint globally and just setting

```
autocmd! BufWritePost * Neomake
let g:neomake_javascript_enabled_makers = ['eslint']
```

results in the same problem.

**Comment by davidosomething** [comment]
ok looking more closely at that, to troubleshoot:

does echoing out `g:neomake_javascript_eslint_exe` in vim ( ` :echo g:neomake_javascript_eslint_exe` ) give you a valid path to an eslint binary?

the code you're using adds the result of `npm bin` to the current path of vim's shell.
it executes that `npm bin` relative to vim's working directory, which may be outside of your project, or in a project where there is no eslint.
If you do `:echo system('npm bin')` from vim, depending on where you are, it may or may not give a valid path to a `node_modules` directory (bad result when package.json is a symlink for example). You need to update `g:neomake_javascript_eslint_exe` to be a valid eslint binary (and looks like you want it to be relative to the buffer you're editing, so you'll need to do it based on the path of the current the buffer)

You can check out my dotfiles for an example of what i've done to use local `node_modules`, you'll need a decent understanding of VimL: https://github.com/davidosomething/dotfiles/blob/master/vim/plugin/plug-neomake.vim#L42 and https://github.com/davidosomething/dotfiles/blob/master/vim/autoload/dkoproject.vim

tl;dr instead of relying on system calls to `npm bin` which are relative vim's working directory, compute the path to eslint based on the path of current buffer and update the buffer-local setting for the eslint_exe using that.

**Comment by adamyonk** [comment]
@davidosomething I always open projects from their git/npm root, so that doesn't bother me too much. I'd like to just get it working and then fix the dynamic paths issues. With these settings:

```
let g:neomake_verbose = 3
autocmd! BufWritePost * Neomake
let g:neomake_javascript_enabled_makers = ['eslint']
let g:neomake_javascript_eslint_exe = system('PATH=$(npm bin):$PATH && which eslint | tr -d "\n"')
```

I see eslint enabled:
`g:neomake_javascript_enabled_makers ['eslint']`

I see a correct path to eslint:
`g:neomake_javascript_eslint_exe  /Users/adamyonk/.code/tesseract/node_modules/.bin/eslint`

I can run said eslint:
```
:!/Users/adamyonk/.code/tesseract/node_modules/.bin/eslint src


/Users/adamyonk/.code/tesseract/src/sagas/requestAvailableTagValues.js
  17:1  error  Parsing error: Unexpected token

✖ 1 problem (1 error, 0 warnings)

shell returned 1

Press ENTER or type command to continue
```

But if I do `:Neomake`:
`Neomake: Nothing to make: no enabled makers`

¯\_(ツ)_/¯

**Comment by adamyonk** [comment]
Oh, and `:Neomake!`:
```
Neomake: Starting: /usr/local/bin/bash -c make
Neomake: makeprg stderr: ["make: *** No targets specified and no makefile found.  Stop.", ""]
Neomake: makeprg stderr done.
Neomake: makeprg processing 1 lines of output
Neomake: makeprg completed with exit code 2
Neomake: All project-level errors cleaned.
Neomake: Removing signs {file: {}, project: {}}
```

**Comment by davidosomething** [comment]
I see -- the docs probably need to be updated to reflect this info:

#### troubleshoot `:Neomake!`

`:Neomake!` runs the default makers defined by `g:neomake_enabled_makers`. I believe that's just the makeprg default, or `make`. `:Neomake!` is used for things like running your project builds, e.g. `gulp build`, so you'd assign that to a new maker and put it into `g:neomake_enabled_makers`

To lint your current buffer, you should run `:Neomake` without the exclamation. This detects the filetype, and uses the makers assigned to that filetype, on the current buffer (e.g. sees current buffer is javascript, sees eslint defined as a maker for js, runs it on buffer)

#### troubleshoot `:Neomake`

What do the following yield, when looking at the buffer:

```
:echo &ft
:echo neomake#GetEnabledMakers(&ft)
:echo executable(g:neomake_javascript_eslint_exe)
```

first one should be `javascript`
second one should be `[ 'eslint' ]`
third one should be `1`

**Comment by adamyonk** [comment]
Aha! Thanks for that explanation, that makes a lot more sense.

OK, so for the :Neomake bit I'm getting:
`javascript.jsx`
`[]`
`1`

Which led me do:
`let g:neomake_jsx_enabled_makers = ['eslint']`

And that got things working with `:Neomake`! But, every save started resulting in:
```
"src/sagas/requestAvailableTagValues.js" 88L, 2176C
Neomake: Starting: /Users/adamyonk/.code/tesseract/node_modules/.bin/eslint -f compact /Users/adamyonk/.code/tesseract/src/sagas/requestAvailableTagValues.js
Press ENTER or type command to continue
```

Can I silence that somehow? So the `BufWritePost` autocmd runs without that prompt?

**Comment by davidosomething** [comment]
you probably still have verbose set?
`let g:neomake_verbose = 3` change to 0

**Comment by adamyonk** [comment]
Gah, yep. It's golden now. Might be worth throwing those troubleshooting steps in the GitHub issue template, that was very helpful to me. Thanks for your help, @davidosomething!

**Comment by blueyed** [comment]
@adamyonk 
> Might be worth throwing those troubleshooting steps in the GitHub issue template

PR? :)

**Comment by blueyed** [comment]
I thought I have understood this issue finally (GetEnabledMakers only returns the union of enabled makers for all filetypes), but `eslint` has been enabled in `jsx` since May 2015 (1aa7420), and also for `javascript`.
So I don't understand while explicitly enabling it for `jsx` helped here..
